### PR TITLE
Surface viewer FMU

### LINKED ISSUE #63: Surface/map comparison viewer
- [ ] Interval thickness maps
- [ ] Difference maps, eg.
    - initial vs. modelled isochore
    - old versus updated structure

### PR DESCRIPTION
**Pull request by HansKallekleiv**
Closes #63.

### CODE REVIEW

File: webviz_subsurface/plugins/_surface_viewer_fmu.py [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+            else:
+                surface = load_surface(self.get_runpath(data, ensemble, real))
+            if real2 in ["Mean", "StdDev", "Min", "Max"]:
+                surface2 = calculate_surface(
+                    self.get_ens_runpath(data2, ensemble2), real2
+                )
+
+            else:
+                surface2 = load_surface(self.get_runpath(data2, ensemble2, real2))
+
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_surface_viewer_fmu.py]: Remove empy lines before `else`, and perhaps instead add one before `if`?

File: webviz_subsurface/plugins/_surface_viewer_fmu.py [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+                if calculation == "Sum":
+                    values = values + surface2.values
+                if calculation == "Product":
+                    values = values * surface2.values
+                if calculation == "Quotient":
+                    values = values / surface2.values
+                if diff_min:
+                    values[values <= diff_min] = diff_min
+                if diff_max:
+                    values[values >= diff_max] = diff_max
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_surface_viewer_fmu.py]: Are some of these mutually exclusive? I.e. `elif`?

File: webviz_subsurface/plugins/_surface_viewer_fmu.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+    if statistic == "StdDev":
+        surface = surfaces.apply(np.nanstd, axis=0)
+    if statistic == "Min":
+        surface = surfaces.apply(np.nanmin, axis=0)
+    if statistic == "Max":
+        surface = surfaces.apply(np.nanmax, axis=0)
+    if statistic == "P10":
+        surface = surfaces.apply(np.nanpercentile, 10, axis=0)
+    if statistic == "P90":
+        surface = surfaces.apply(np.nanpercentile, 90, axis=0)
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_surface_viewer_fmu.py]: `elif`?

File: webviz_subsurface/plugins/_surface_viewer_fmu.py [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+
+
+@CACHE.memoize(timeout=CACHE.TIMEOUT)
+def get_surfaces(fns):
+    return xtgeo.surface.surfaces.Surfaces(fns)
+
+
+@webvizstore
+def get_path(path) -> Path:
+
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_surface_viewer_fmu.py]: Remove empty line.

File: webviz_subsurface/plugins/_surface_viewer_fmu.py [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+        json.dumps(
+            sorted([str(filename) for filename in folder.glob(f"*{suffix}")])
+        ).encode()
+    )
+
+
+def make_fmu_filename(data):
+    if data["date"] is None:
+        return f"{data['name']}--{data['attr']}"
+    return f"{data['name']}--{data['attr']}--{data['date']}"
```
Comment (anders-kiaer) [inline webviz_subsurface/plugins/_surface_viewer_fmu.py]: Consider
```python
    filename = f"{data['name']}--{data['attr']}"

    if data["date"] is None:
        filename += f"--{data['date']}"

    return filename 
```

File: webviz_subsurface/_datainput/fmu_input.py [hunk 6]
```diff
@@ -114,20 +115,31 @@ def find_surfaces(ensemble_paths: dict, suffix="*.gri", delimiter="--"):
     files = []
     for _, path in ensemble_paths.items():
         path = Path(path)
-        files += glob.glob(str(path / "share" / "results" / "maps" / suffix))
+        for realpath in glob.glob(str(path / "share" / "results" / "maps" / suffix)):
```
[the remark this answers is not in the document]
Comment (HansKallekleiv) [reply webviz_subsurface/_datainput/fmu_input.py]: The challenge is the two wildcards:
`/scratch/case/realization-*/iter-0/share/results/maps/*.gri` 
With `Path().glob()` I think we would have to do something like:

`(path).glob(f"realization-*/{ensemble}/share/results/maps/*.gri")`

Not sure what is cleanest?

[the remark this answers is not in the document]
Comment (anders-kiaer) [reply webviz_subsurface/_datainput/fmu_input.py]: Ah true, then I find `glob.glob` to be the cleanest 👍

File: webviz_subsurface/_private_plugins/surface_selector.py [hunk 7]
```diff
@@ -72,73 +76,29 @@ def set_ids(self):
...
-        self.realization_id_btn_next = f"{uuid}-realization-btn-next"
         self.name_wrapper_id = f"{uuid}-name-wrapper"
         self.date_wrapper_id = f"{uuid}-date-wrapper"
-        self.ens_wrapper_id = f"{uuid}-ens-wrapper"
-        self.real_wrapper_id = f"{uuid}-real-wrapper"
-        self.aggreal_id = f"{uuid}-aggreal"
-        self.sens_name_id = f"{uuid}-sens-name-id"
-        self.sens_case_id = f"{uuid}-sens-case-id"
-        self.sens_name_wrapper_id = f"{uuid}-sens-name-wrapper-id"
-        self.sens_case_wrapper_id = f"{uuid}-sens-case-wrapper-id"
```
[the remark this answers is not in the document]
Comment (anders-kiaer) [reply webviz_subsurface/_private_plugins/surface_selector.py]: After discussion, we will leave as is, due to it being a "private plugin".

**Comment by anders-kiaer** [comment]
LGTM! 🙂 Remember to update `.travis.yml` first.

**Pull request merged**